### PR TITLE
[CPU Backend] Vectorized top-k and top-p sampling kernels

### PR DESCRIPTION
**Pull request by tianmu-li**
## Purpose

Adds SIMD-vectorized C++ kernels for top-k and top-p logit filtering on CPU, replacing the previous `apply_top_k_top_p_pytorch` fallback that sorted the full vocabulary on every step.

**Key changes:**

- **`csrc/cpu/topk_topp_sampler.cpp`** (new): AVX-512/AVX2 C++ kernels for top-k, top-p, and fused top-k+top-p. Scalar fallback for ARM/POWER/RISC-V.
- **`csrc/cpu/cpu_arch_macros.h`**: Adds the AVX2 `DEFINE_FAST_EXP` macro for vectorized softmax on AVX2.
- **`cmake/cpu_extension.cmake`**: Wires the new file into all three build targets; adds `-mfma` for AVX2.
- **`csrc/cpu/torch_bindings.cpp`** + **`vllm/_custom_ops.py`**: Registers `cpu_topp_sampling`, `cpu_topk_sampling`, `cpu_topk_topp_sampling` ops under the `_C` namespace.
- **`vllm/v1/sample/ops/topk_topp_sampler.py`**: New `apply_top_k_top_p_cpu()` dispatch function; falls back to PyTorch-native when C++ ops are unavailable.
- **`tests/v1/sample/test_topk_topp_sampler.py`**: Adds `TestCpuTopkTopp` with 17 tests covering all filter combinations, edge cases, and fused-vs-sequential consistency.

The algorithm is inspired by `topk_topp_triton.py` ([Qrita paper](https://arxiv.org/abs/2602.01518)): same outlier-gather fast path, ternary search for the top-k pivot, lookup tables, 18-iteration limit, and tie-breaking tolerances.

## Test Plan
Unit test:
```bash
python benchmarks/kernels/cpu/benchmark_cpu_topk_topp.py --vocab-sizes 131072 --quiet
```
E2E test:
Server (vllm bench throughput uses top_p=1, which won't invoke top-p sampler):
```
MODEL_NAME=RedHatAI/Meta-Llama-3.1-8B-Instruct-quantized.w8a8
vllm serve $MODEL_NAME \
    --served-model-name meta-llama/Llama-3.1-8B-Instruct \
	--dtype bfloat16 \
	--tensor-parallel-size=4
```

Client
```
MODEL_NAME=meta-llama/Llama-3.1-8B-Instruct
vllm bench serve --model $MODEL_NAME \
 --num-prompt 128 \
 --random-input-len {128,2048} \
 --random-output-len {128, 2048}
```

## Test Result
Unit:
| Scenario | Batch | CPU (ms) | PyTorch (ms) | Speedup |
|---|---:|---:|---:|---:|
| top-k only, whole batch (k=50) | 1 | 0.312 | 0.439 | 1.4× |
| top-k only, whole batch | 64 | 0.641 | 1.706 | 2.7× |
| top-k only, whole batch | 512 | 4.806 | 20.249 | 4.2× |
| top-k only, whole batch | 2048 | 18.214 | 66.785 | 3.7× |
| top-p only, whole batch (p=0.9) | 1 | 0.717 | 10.047 | 14.0× |
| top-p only, whole batch | 64 | 1.345 | 28.695 | 21.3× |
| top-p only, whole batch | 512 | 9.512 | 212.159 | 22.3× |
| top-p only, whole batch | 2048 | 34.791 | 781.781 | 22.5× |
| top-k + top-p, whole batch | 64 | 1.201 | 28.527 | 23.8× |
| top-k + top-p, whole batch | 512 | 8.631 | 232.781 | 27.0× |
| top-k + top-p, whole batch | 2048 | 31.257 | 830.235 | 26.6× |
| mixed partial (1/3 k, 1/3 p, 1/3 both) | 512 | 6.683 | 230.895 | 34.6× |
| mixed partial | 2048 | 24.586 | 853.446 | 34.7× |

E2E:
| ISL/OSL | Pre | Post | Speedup |
| ------- | --- | ---- | ------- |
| 128/128 | 1032 | 1527 | 1.47x |
| 2048/128 | 271 | 287 | 1.06x |
| 128/2048 | 1014 | 1450 | 1.43x |

**AI assistance**

This implementation was developed with Claude Code assistance. All changed lines were reviewed by the human author, tests were run locally, and the logic was verified end-to-end including error-path probing.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by bigPYJ1151 — COMMENTED** [review]
Thanks @tianmu-li :) 
I have enabled Triton for CPU in #43225 and pushing for pre-built Triton CPU wheel release. Triton has been enabled by default for vLLM x86 CPU images, and will be enabled by default for x86 wheels in the near future. 
So could you help to check whether it is a better option to use the Triton implmentation?

**Comment by tianmu-li** [comment]
Hi @bigPYJ1151. Thanks for bringing this up. I created another PR using Triton instead #43633. Given that performance is better with Triton, I think this PR can be closed if we choose to use Triton.
